New Scenario Step: SAP UI5 on SAP Cloud Platform (#388)
SAP/jenkins-library: documentation/docs/scenarios/ui5-sap-cp/Readme.md

# Docu stub

# Project template files

> The following template files needs to be provided on project level:

Reviewer: ```suggestion
The following template files needs to be provided and adjusted on project level:
```

# Pipeline for SAP UI5/Fiori On SAP Cloud Platform

This is a so called scenario step. Scenario steps are aggregations of several steps implementing a simple, but complete pipeline. This should make simple scenarios easy to set up. Your `Jenkinsfile` can be as simple as:

```groovy
@Library('piper-lib-os') _

<scenario>Pipeline script: this
```

## Description

This steps builds an SAP UI5 or Fiori based application using MTA and deploys the build result into an SAP Cloud Platform (Neo) account. This scenario wraps the [mtaBuild](../../steps/mtaBuild.md) and [neoDeploy](../../steps/neoDeploy.md) steps.

![This pipeline in Jenkins Blue Ocean](images/pipeline.jpg)

## Prerequisites

### General Prerequisites

- project "Piper" requires a Jenkins (2.x) with pipeline plugins to run

> - project "Piper" needs to be [registered in Jenkins](https://github.com/SAP/jenkins-library/blob/master/README.md) as a Global Shared Library

Reviewer: ```suggestion
- project "Piper" needs to be [registered in Jenkins](https://github.com/SAP/jenkins-library/blob/master/README.md#download-and-installation) as a Global Shared Library
```

> # Docu stub

Reviewer: ```suggestion
# FioriOnCloudPlatformPipeline
```

In general I'd follow the usual pattern for step documentation.

> # Purpose

Reviewer: ```suggestion
# Description
```

> Here we should describe what this scenario does: It builds a SAPUI5 based application and deploys

Reviewer: ```suggestion
This pipeline steps builds an SAP UI5 based application using MTA and deploys
```

> the build result into a NEO account. Build is performed via mta -> node -> grunt -> sap-best-practices plugin

Reviewer: ```suggestion
the build result into an SAP Cloud Platform (Neo) account. This scenario wraps the [mtaBuild](mtaBuild.md) and [neoDeploy](neoDeploy.mta) steps.
```

> In fact it is `grunt clean, build, lint`.

Reviewer: We can leave that out, it's in the mtaBuild doc.

> # The following tools are used

Reviewer: I'd skip this section:
this comes from prerequisites

 - mta (jar file which must be present on the build server; later: via docker)
 - npm (needs to be installed and configured on the build server: later: via docker)
 - grunt (will be materialized through npm)
 - sap best practices (will be materialized through npm)

> # Prerequisites (in the central environment)

Reviewer: I would reference to the requirements of the underlying steps, i.e. mtaBuild and neoDeploy + list/link generic prerequisites

- project "Piper" requires a Jenkins (2.x) with pipeline plugins to run
- project "Piper" needs to be [registered in Jenkins](https://github.com/SAP/jenkins-library/blob/master/README.md) as a Global Shared Library
- prerequisites for the [MTA build](mtaBuild.md)
- prerequisites for the [deployment](neoDeploy.md)

  - piper lib registered
  - npm installed and configured so that
      - Grunt and the
      - sap best practices build plugin can be materialized
  - Neo deploy account available, credentials maintained in Jenkins.

> # The reader should already know

Reviewer: Is this relevant? It's in prerequisites actually.

  - what a shared lib is and how it can be registered.
  - understanding of node /grunt build
  - SAP WEBIDE and the corresponding development workflow.

> # Workflow

Reviewer: Not sure this is relevant. I'd skip this section.

  - Developer works inside WEB-IDE and commits changes into this git clone.
  - Afterwards s?he pushes into a shared git repo.
  - Jenkins (push or pull?) detects the changes and triggers a build.
  - The build result gets deployed into the configured NEO account.

# References
- https://developers.sap.com/germany/tutorials/webide-grunt-basic.html

# Open questions:

>   * What is our relationship to the SAP-WEB-IDE?

Reviewer: none

  In this case it needs to be ensured that this dependency is available during local development as well as in central
  build infrastructure. Beside this the might might of course contain other npm configuration settings.
- *(mta.yaml)[documentation/docs/scenarios/ui5-sap-cp/files/mta.yaml]* Controls the behavior of the mta toolset. Placeholders
  (labeled like this `<placeholder`) needs to be replaced by valid values (version, applicationName). The `${timestamp}` in the version is replaced
  by the piperLibrary step `mtaBuild` in order to be able to distiguish the build results. <comment mh>I'm not happy with
  that approach modifying the sources here, we should check if this can be done in a better way.</comment mhend>
- *(package.json)[documentation/docs/scenarios/ui5-sap-cp/files/package.json]* Must contain a development dependency to `@sap/grunt-sapui5-bestpractice-build`. And of course other dependencies if required.
  Name, version and description needs to be maintained, too (placeholders).
- (Guntfile.js)[documentation/docs/scenarios/ui5-sap-cp/files/Gruntfile.js] controls the grunt build. By default these tasks are executed: `clean`, `build`, `lint`. It is

>   less likely that this file needs to be changed.

Reviewer: I'd add all this into the docs of mtaBuild and/or neoDeploy and link.
